### PR TITLE
Make fixes related to using Mbed Crypto as a service

### PR DESCRIPTION
**Pull request by Patater**
An eclectic grab bag of hot commits!

Make fixes needed for using Mbed Crypto as a service (as on a v8-M platform). Also update the getting started guide with working code snippets and text.

### CODE REVIEW

File: include/psa/crypto_types.h [hunk 1]
```diff
@@ -120,6 +120,7 @@ typedef uint32_t psa_key_lifetime_t;
  * psa_key_id_t in crypto_platform.h instead of here. */
 #if !defined(MBEDTLS_PSA_CRYPTO_KEY_FILE_ID_ENCODES_OWNER)
 typedef uint32_t psa_key_id_t;
+#define PSA_KEY_ID_INIT 0
```
Comment (yanesca) [inline include/psa/crypto_types.h]: Why do we need this?

File: include/psa/crypto_struct.h [hunk 2]
```diff
@@ -422,7 +422,14 @@ static inline void psa_set_key_lifetime(psa_key_attributes_t *attributes,
 {
     attributes->core.lifetime = lifetime;
     if( lifetime == PSA_KEY_LIFETIME_VOLATILE )
+    {
+#ifdef MBEDTLS_PSA_CRYPTO_KEY_FILE_ID_ENCODES_OWNER
+        attributes->core.id.key_id = 0;
```
Comment (yanesca) [inline include/psa/crypto_struct.h]: Could we use the macro initialiser here?

Comment (Patater) [reply include/psa/crypto_struct.h]: No, because this is assignment not initialization.

File: include/psa/crypto_types.h (hunk 1 above)
Comment (Patater) [reply include/psa/crypto_types.h]: I was hoping to be able to use it in the assignment later on, without having to add an ifdef for MBEDTLS_PSA_CRYPTO_KEY_FILE_ID_ENCODES_OWNER in psa_set_key_lifetime. That didn't pan out.

File: docs/getting_started.md [hunk 3]
```diff
@@ -357,18 +361,19 @@ At this point the derived key slot holds a new 128-bit AES-CTR encryption key de
...
 
     /* Import a key for use in key derivation, if such a key has already been imported you can skip this part */
-    psa_key_policy_set_usage(&policy, PSA_KEY_USAGE_DERIVE, alg);
-    status = psa_set_key_policy(base_key, &policy);
+    psa_key_set_usage_flags(&attributes, PSA_KEY_USAGE_DERIVE);
+    psa_key_set_algorithm(&attributes, alg);
+    psa_key_set_type(&attributes, PSA_KEY_TYPE_DERIVE);
 
-    status = psa_import_key(base_key, PSA_KEY_TYPE_DERIVE, key, sizeof(key));
+    XXX needs work XXX
```
[the remark this answers is not in the document]
Comment (Patater) [reply docs/getting_started.md]: The getting started guide needs updating for the new way to derive keys.

File: include/psa/crypto_types.h (hunk 1 above)
Comment (yanesca) [reply include/psa/crypto_types.h]: Is there any reason why we shouldn't remove it?

Comment (Patater) [reply include/psa/crypto_types.h]: I did end up using it in the initialization macros for key attributes.

Comment (yanesca) [reply include/psa/crypto_types.h]: But we have it in `crypto_platform.h` already, is there any reason why we would want to maintain both definitions?

Comment (Patater) [reply include/psa/crypto_types.h]: I've removed the extra one from `crypto_platform.h`. Now, we have the INIT macro next to the typedef in all cases.

**Comment by Patater** [comment]
Rebased to remove extra `PSA_KEY_ID_INIT` from `crypto_platform.h` in commit "psa: Add PSA_KEY_ID_INIT".

File: docs/getting_started.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Patater) [reply docs/getting_started.md]: I've added code for the new API, but need to update the text everywhere still. Turns out the guide was more out of date than expected: not just key derivation.

**Comment by Patater** [comment]
Rebased to update the getting started document to latest API, with all snippets tested as per https://github.com/ARMmbed/mbed-os-example-mbed-crypto/pull/40

File: docs/getting_started.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Patater) [reply docs/getting_started.md]: I've completed the update to the whole guide now.

**Comment by Patater** [comment]
CI failures are only API changes which are intentional: change type of key from key_id to app_key_id

File: docs/getting_started.md [hunk 4]
```diff
@@ -63,84 +63,121 @@ To use the Mbed Crypto APIs, call `psa_crypto_init()` before calling any other A
...
-
-To sign a given message `payload` using RSA:
-1. Set the key policy of the chosen key slot by calling `psa_key_policy_set_usage()` with the `PSA_KEY_USAGE_SIGN` parameter and the algorithm `PSA_ALG_RSA_PKCS1V15_SIGN_RAW`.
-This allows the key in the key slot to be used for RSA signing.
-1. Import the key into the key slot by calling `psa_import_key()`. You can use an already imported key instead of importing a new one.
-1. Call `psa_asymmetric_sign()` and get the output buffer that contains the signature:
+* Have a valid key with appropriate attributes set:
+    * Usage flag `PSA_KEY_USAGE_SIGN` to allow signing.
+    * Usage flag `PSA_KEY_USAGE_VERIFY` to allow signature verification.
+    * Algorithm set to desired signature algorithm
```
Comment (AndrzejKurek) [inline docs/getting_started.md]:
```suggestion
    * Algorithm set to desired signature algorithm.
```

File: docs/getting_started.md [hunk 5]
```diff
@@ -156,32 +193,78 @@ Encrypting a message with a symmetric cipher:
 1. Call `psa_cipher_update` one or more times, passing either the whole or only a fragment of the message each time.
 1. Call `psa_cipher_finish` to end the operation and output the encrypted message.
 
-Encrypting random data using an AES key in cipher block chain (CBC) mode with no padding (assuming all prerequisites have been fulfilled):
+Encrypting some data using an AES key in cipher block chain (CBC) mode with no padding (assuming all prerequisites have been fulfilled):
```
Comment (AndrzejKurek) [inline docs/getting_started.md]: "Some" sounds a bit funny to me, especially in docs - maybe it should just be "Encrypt data" (here and in the source code)?

File: docs/getting_started.md [hunk 6]
```diff
@@ -304,86 +488,172 @@ Multiple sequential calls to `psa_hash_abort` on an operation that has already b
...
+    * Usage flags set for key derivation (`PSA_KEY_USAGE_DERIVE`)
+    * Key type set to `PSA_KEY_TYPE_DERIVE`.
+    * Algorithm set to a key derivation algorithm
+      (`PSA_ALG_HKDF(PSA_ALG_SHA_256)`).
+
+Deriving a new AES-CTR 128-bit encryption key into a given key slot using HKDF
+with a given key, salt and info:
+1. Set up the key derivation context using the `psa_key_derivation_setup`
+function, specifying the derivation algorithm `PSA_ALG_HKDF(PSA_ALG_SHA_256)`.
+1. Provide an optional salt with `psa_key_derivation_input_bytes`
```
Comment (AndrzejKurek) [inline docs/getting_started.md]: Previously each of these lines ended with a period.

File: docs/getting_started.md [hunk 7]
```diff
@@ -304,86 +488,172 @@ Multiple sequential calls to `psa_hash_abort` on an operation that has already b
...
+        return;
+    }
+    status = psa_key_derivation_input_bytes(&operation,
+                                            PSA_KEY_DERIVATION_INPUT_INFO,
+                                            info, sizeof(info));
+    if (status != PSA_SUCCESS) {
+        printf("Failed to input info (expand)\n");
+        return;
+    }
+    psa_set_key_usage_flags(&attributes, PSA_KEY_USAGE_ENCRYPT);
```
Comment (AndrzejKurek) [inline docs/getting_started.md]: Can attributes be reused here? Is resetting them equal to having `PSA_KEY_ATTRIBUTES_INIT`?

File: docs/getting_started.md (hunk 4 above)
Comment (Patater) [reply docs/getting_started.md]: I didn't want to put a period at the end if the line wasn't a compete sentence. I'll change this to be consistent at least.

File: docs/getting_started.md (hunk 7 above)
Comment (Patater) [reply docs/getting_started.md]: Yes.

**Comment by Patater** [comment]
Rebased to address feedback

**Pull request merged**